c++/src/kj/compat/tls.c++

```diff
@@ -57,10 +57,20 @@ void throwOpensslError() {
   // throw it.
 
   kj::Vector<kj::String> lines;
-  while (unsigned long long error = ERR_get_error()) {
+  const char *fname;
+  int lno;
+  const char *data;
+  int flags;
+  while (unsigned long long error = ERR_get_error_line_data(&fname, &lno, &data, &flags)) {
     char message[1024];
     ERR_error_string_n(error, message, sizeof(message));
-    lines.add(kj::heapString(message));
+
+    bool has_data = (flags & ERR_TXT_STRING) != 0;
+    const char *extra_prefix = has_data ? " [" : "";
+    const char *extra_suffix = has_data ? "]" : "";
+    const char *extra = has_data ? data : "";
+
+    lines.add(kj::str("[", fname, ":", lno, "] ", message, extra_prefix, extra, extra_suffix));
```

Comment:
I take it `fname` is the OpenSSL source file name?

`kj::Exception` itself already stores a filename and line number, maybe we should initialize that directly? Like so:

```
kj::throwFatalException(kj::Exception(kj::Exception::Type::FAILED,
    fname, lno, kj::str(message, extra_prefix, extra, extra_suffix));
```

The one problem is how to deal with multiple errors. In that case I'd say we want the overall exception's source location to match either the first or last error (not sure which is more appropriate...) and then the rest could be in the description... maybe?

Reply:
> I take it fname is the OpenSSL source file name?

That's correct, yep. Sometimes it's a weird relative path like `../ssl/ssl_lib.c`, seems to depend on how OpenSSL was built.

> The one problem is how to deal with multiple errors. In that case I'd say we want the overall exception's source location to match either the first or last error (not sure which is more appropriate...) and then the rest could be in the description... maybe?

Ah I think the original semantics make sense, like I'd expect the filename + lineno to match the exception "callsite" myself. If including filename + line number info from OpenSSL in the error message makes you uneasy I'm not particularly attached to it. The `data` fragment is more important IMO.

Reply:
Note the exception's stack trace will already identify the location where it is constructed, so identifying the location of the KJ assertion is actually redundant.

To me, what we're doing here is translating an openssl exception to a KJ exception, and it feels right for the filename / line number to be part of that translation.

But on a practical note, I guess the OpenSSL source location probably isn't that useful to anyone who isn't an OpenSSL developer, so maybe that's a good argument for just omitting it?